nixos/modules/services/web-apps/pixelfed.nix

```diff
@@ -0,0 +1,472 @@
+      phpPackage = mkPackageOptionMD pkgs "php81" { };
+
+      user = mkOption {
+        type = types.str;
+        default = "pixelfed";
+        description = lib.mdDoc ''
+          User account under which pixelfed runs.
+
+          ::: {.note}
+          If left as the default value this user will automatically be created
+          on system activation, otherwise you are responsible for
+          ensuring the user exists before the pixelfed application starts.
+          :::
+        '';
+      };
+
+      group = mkOption {
+        type = types.str;
+        default = "pixelfed";
+        description = lib.mdDoc ''
+          Group account under which pixelfed runs.
+
+          ::: {.note}
+          If left as the default value this group will automatically be created
+          on system activation, otherwise you are responsible for
+          ensuring the group exists before the pixelfed application starts.
+          :::
+        '';
+      };
+
+      localDomain = mkOption {
+        type = types.str;
+        description = lib.mdDoc ''
+          FQDN for the Pixelfed instance.
+        '';
+      };
+
+      secretFile = mkOption {
+        type = types.path;
+        description = lib.mdDoc ''
+          A secret file to be sourced for the .env settings.
+          Put `APP_KEY` and other settings that should not end up in the Nix store.
+        '';
+      };
+
+      settings = mkOption {
+        type = with types; (attrsOf (oneOf [ bool int str ]));
+        description = lib.mdDoc ''
+          .env settings for Pixelfed.
+          Secrets should use `secretFile` option.
+        '';
+      };
+
+      nginx = mkOption {
+        type = types.nullOr (types.submodule
+          (import ../web-servers/nginx/vhost-options.nix {
+            inherit config lib;
+          }));
+        default = null;
+        example = lib.literalExpression ''
+          {
+            serverAliases = [
+              "pics.''${config.networking.domain}"
+            ];
+            enableACME = true;
+          }
+        '';
+        description = lib.mdDoc ''
+          With this option, you can customize an nginx virtual host which already has sensible defaults for Dolibarr.
+          Set to {} if you do not need any customization to the virtual host.
+          If enabled, then by default, the {option}`serverName` is
+          `''${domain}`,
+          If this is set to null (the default), no nginx virtualHost will be configured.
+        '';
+      };
+
+      redis.createLocally = mkEnableOption
+        (lib.mdDoc "a local Redis database using UNIX socket authentication")
+        // {
+          default = true;
+        };
+
+      database = {
+        createLocally = mkEnableOption
+          (lib.mdDoc "a local database using UNIX socket authentication") // {
+            default = true;
+          };
+        automaticMigrations = mkEnableOption
+          (lib.mdDoc "automatic migrations for database schema and data");
```

Comment:
What happens if this is set to false?

Reply:
Subsequent updates of the server won't run the migrations automatically, the server will potentially fail because of a migration check and an admin has to intervene to do migrations reading the required release notes.

Reply:
Most other service just do the migrations on startup and do not have such a setting. Not sure if we need it here at all.

Reply:
I can turn it to on by default but I don't agree that other services are good examples of this behavior.

Reply:
Why not? I would want my service to migrate it's DB on an upgrade and not to break.

Reply:
Because it is trusting that either the end user has proper backups/snapshots to rollback properly the data in case of failure and that the developers did not fuck up. We have prior art on programs doing migrations and erasing all the data as a result, see Hedgedocs history in nixpkgs.

Auto migration is a conscious choice by administrators that should probably accompanied by measures or explicit risk taking based on the reputation of the software in question.

Reply:
(and I am fine with a "nixos.enableAutoMigrations = true;" which could be global and be a general policy choice, but I am not okay with the statut quo.)

Reply:
> thats something for an RFC

sure

> sure it is not perfect but doing this for one service just creates confusion and makes this one service work rather different from all other services we have.

uncertain that all services are doing it IIRC, I experienced some services where there was no auto migrations.

Anyway, I can settle for enabling it auto.